src/Microsoft.OData.Core/UriParser/SemanticAst/EachSegment.cs

```diff
@@ -0,0 +1,90 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="EachSegment.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.UriParser
+{
+    #region Namespaces
+    using Microsoft.OData.Edm;
+    #endregion Namespaces
+
+    /// <summary>
+    /// A segment representing $each in a path.
+    /// </summary>
+    public sealed class EachSegment : ODataPathSegment
+    {
+        /// <summary>
+        /// The <see cref="IEdmType"/> of the resource that this <see cref="EachSegment"/> represents.
+        /// </summary>
+        private readonly IEdmType edmType;
+
+        /// <summary>
+        /// Build a segment representing each.
+        /// </summary>
+        /// <param name="navigationSource">The entity collection that this set-based operation applies to.</param>
+        /// <param name="targetEdmType">Target type for the entity being referenced.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if any input parameter is null.</exception>
+        /// <remarks>$filter cannot be applied on singletons.</remarks>
+        public EachSegment(IEdmNavigationSource navigationSource, IEdmType targetEdmType)
+        {
+            ExceptionUtils.CheckArgumentNotNull(navigationSource, "navigationSource");
+
+            this.Identifier = UriQueryConstants.EachSegment;
+            this.SingleResult = false;
+            this.TargetEdmNavigationSource = navigationSource;
+            this.TargetKind = RequestTargetKind.Resource;
+            this.TargetEdmType = targetEdmType ?? navigationSource.EntityType();

```

Comment:
>navigationSource.EntityType() [](start = 50, length = 29)

why navigationSource.EntityType() rather than navigationSource.Type?

Reply:
navigationSource.EntityType() helps resolve to the target EdmType while navigationSource.Type resolves to the collection of the type.

navigationSource.EntityType():
TargetEdmType	{Fully.Qualified.Namespace.Person}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.Csdl.CsdlSemantics.CsdlSemanticsEntityTypeDefinition}

navigationSource.Type:
Type	{Collection([Fully.Qualified.Namespace.Person Nullable=False])}	Microsoft.OData.Edm.IEdmType {Microsoft.OData.Edm.EdmCollectionType}